### ISSUE TITLE
Wrong error message for ambiguous given instances

### ISSUE THREAD

**Issue by KuceraMartin** [scala/scala3#19414]
## Compiler version
3.3.1, 3.4.0-RC1-bin-20240109-91db06a-NIGHTLY

## Minimized code
```scala
trait JsValue
trait JsObject extends JsValue

trait JsonWriter[T]
trait BodySerializer[-B]

class JsonPrinter

given JsonWriter[JsValue] = ???
given JsonWriter[JsObject] = ???

given[B : JsonWriter](using printer: JsonPrinter = new JsonPrinter): BodySerializer[B] = ???

def f: Unit =
  summon[BodySerializer[JsObject]]
```

## Output
```
[error] No given instance of type BodySerializer[JsObject] was found for parameter x of method summon in object Predef.
[error] I found:
[error]
[error]     given_BodySerializer_B[B]()
[error]
[error] But given instance given_BodySerializer_B does not match type BodySerializer[JsObject].
[error]   summon[BodySerializer[JsObject]]
[error]                                   ^
```

## Expectation
```
[error] Ambiguous given instances: both given instance given_JsonWriter_JsValue and given instance given_JsonWriter_JsObject match type JsonWriter[B] of parameter x of method summon in object Predef
[error]   summon[BodySerializer[JsObject]]
[error]                                   ^
```

note: the expectation is what I got when I removed the implicit `JsonPrinter` parameter from the given `BodySerializer`:
```scala
given[B : JsonWriter]: BodySerializer[B] = ???
```

**Comment by joroKr21** [comment]
Perhaps a related issue - when the instance below `I found` is a macro, it might have a custom error message.
It would be nice to show the custom error message instead of a type mismatch.

**Comment by mbovel** [comment]
I add a `given JsonPrinter = new JsonPrinter`, I also get the expected error message.

**Comment by mbovel** [comment]
Note: it works if the default parameter is in a separate list:

```scala
given[B](using writer: JsonWriter[B])(using printer: JsonPrinter = new JsonPrinter): BodySerializer[B] = ???
```

**Comment by Jezisek-Ematiq** [comment]
The contextual bound desugars to an implicit rather than using which seems to cause a problem in Applications.scala:641 (it doesn't fall into the else if branch because it's not a contextual function)

**Comment by mbovel** [comment]
I don't understand why `propagatedFailure` has its current semantic:

https://github.com/lampepfl/dotty/blob/641ab7a4984cd76ff77fadaf822472a0a930e4a2/compiler/src/dotty/tools/dotc/typer/Typer.scala#L3853-L3867

It returns the first non-ambiguous error if there is one, or else the first error, or else `NoType`.

Written otherwise:

```scala
val firstNonAmbiguous = args.tpes.find(!_.isInstanceOf[AmbiguousImplicits])
def firstError = args.tpes.find(_.isError)
val propFail = firstNonAmbiguous.orElse(firstError).getOrElse(NoType)
```

Why isn't it the opposite; why doesn't it return the first ambiguous error if there is one?

```scala
val firstAmbiguous = args.tpes.find(_.isInstanceOf[AmbiguousImplicits])
def firstError = args.tpes.find(_.isError)
val propFail = firstAmbiguous.orElse(firstError).getOrElse(NoType)
```

Can adding default arguments avoid an ambiguous error? That would explain it. If note, we could probably report it directly.

**Comment by odersky** [comment]
>  don't understand why propagatedFailure has its current semantic:

Ambiguous implicits often arise when something else is failing (for instance a variable was not instantiated enough). That's why we try to find first other errors. That was the thinking behind the current semantic. So I am not sure reversing this as in #19647 is always the thing to do. It might make error messages worse in other contexts.

**Comment by mbovel** [comment]
Looking again into this, the main difference between the case with the default arg and the case without is that in the former, `issueErrors` is not called, so the tree doesn't get a `SearchFailure` type and the failure doesn't not "bubble up", while it does in the other case.

I made two tentatives to call it in the default args case:

1. https://github.com/scala/scala3/commit/292865af3f43e11783ff99629f6846c788069a1f: Parametrize `issueErrors` with the argument trees and call it with the retyped arg trees in the default args case. However, the retyped tree is not always an `Apply` node.
2. https://github.com/scala/scala3/commit/7468f2594f06917d1fa33628e2c0bda143093030: Call `issueErrors` when the type of the retyped tree in the default args case is an error, but use the arguments of the apply tree without default args and filter out default arguments.

Neither look very nice… Not sure if there is a simpler way.

By the way, I also found why the parameters name is wrong: as the `SearchFailure` bubble up, `issueFailure` can't know if the failure comes from a nested application or not. Currently, it just assumes it's not nested, which is wrong for the test case of this issue.
